[Quantum] Fix handling of Sum in QExpr and qapply

ellisonbg opened the pull request:
#### References to other Issues or PRs

Fixes #26495 

Fixes #23820

#### Brief description of what is fixed or changed

This PR fixes some bugs in the handling of sums in quantum expressions. Previously, quantum expressions (subclasses of `QExpr` didn't have a proper `free_symbols` property. Thus, when used inside `Sum`, the `QExpr` instances were being pulled outside the `Sum`. This PR adds the proper `free_symbols` property and along with tests for `QExpr`. In addition, now that this works properly, `qapply` needed to be updated to handle `Sum`. This support was added, without any major architectural changes to how `qapply` works.

#### Other comments

Refactoring of `qapply` will be done in future work, this PR keeps the overall approach the same.

Changes should be non-breaking.

A new keyword argument to `qapply` has been introduced, `sum_doit`, that will call `.doit()` on any sums that are encountered in `qapply`. This defaults to `False` to avoid breaking existing behavior.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* physics.quantum
  * All quantum expressions (operators and states) now have a proper free_symbols property, which is needed for them to work with Sum.
  * qapply now handles Sum instances in quantum expressions.
<!-- END RELEASE NOTES -->

ellisonbg committed:
quantum: Fix free_symbols on QExpr to enable Sums over quantum expressions

At some point, the free_symbols property of QExpr was changed from the base implementation of Basic/Expr. This was breaking Sums of quantum expressions. The solution is to remove the overridden version and rely on the base classes. Tests added and clarified. See issues #26495 and #23820.

ellisonbg committed:
quantum: Removed overridden free_symbols on State

Now that QExpr is no longer overriding free_symbols, State can go back to using the base implementation implicitely.

ellisonbg committed:
quantum: Added logic in qapply to handle Sum.

Previously, quantum expressions with Sums were not being handled. This change introduces full handling for Sums, both at the top level and in the qapply_Mul section. The implementation brings the sums to the left of products and calls qapply_Mul on the function of the Sum. To keep backwards compatibiilty, I have introduced a new sum_doit keyword argument to qapply that defaults to False. When set to True, this calls Sum.doit() on any Sums found in the process of applying operators to states.

Note: qapply does need some refactoring. After these changes I have some ideas about that, but I don't want to mix those improvements with this fix.

ellisonbg committed:
quantum: Fix set literal in test_qexpr

This fixes a minor issue for ruff.

oscarbenjamin commented on sympy/physics/quantum/qexpr.py:
```diff
@@ -97,9 +97,9 @@ class QExpr(Expr):
-    @property
-    def free_symbols(self):
-        return {self}
+    # @property
+    # def free_symbols(self):
+    #     return {self}
```
It is good to remove these things. What does this method even do?

ellisonbg replied:
Opps, forgot to actually remove these lines. I have no idea why the free_symbols method was written to return `{self}`. The PR that did that didn't have any useful comments about it. New commit removes the lines as intended.

ellisonbg committed:
quantum: Removing the free_symbols method

In the first pass, I commented out the bad free_symbols method rather than removing it. This commit does the actual removal.

oscarbenjamin commented on sympy/physics/quantum/qapply.py:
```diff
@@ -177,18 +190,40 @@ def qapply_Mul(e, **options):
+
     # Now try to actually apply the operator and build an inner product.
     try:
         result = lhs._apply_operator(rhs, **options)
-    except NotImplementedError:
+    except (NotImplementedError, AttributeError):
         result = None
```
In general I think it is better if we don't catch exceptions like `AttributeError`. It can easily mask bugs and other things and is not really a good way of doing control flow.

What is this `AttributeError` being raised by? If the exception is being raised then does that not really indicate a bug elsewhere like some type is missing a method or something?

ellisonbg replied:
Fixed in a commit this morning. Thanks!

ellisonbg committed:
quantum: Improve error handling in qapply for Operator cases

This narrows the exception handling in qapply to not be overly broad in catching AttributeError.

ellisonbg commented:
Looks like tests are passing and review comments addressed. Let me know if there is anything else that needs to be addressed.

oscarbenjamin commented:
I can't really review the mathematical details but the code looks good: I'll trust you on the quantum mechanics.

Thanks.
